### ETL Project - The Best 50 Superhero Movies

ETL is a type of data integration that refers to the three steps (extract, transform, load) used to blend data from multiple sources. It's often used to build a data warehouse. During this process, data is taken (extracted) from a source system, converted (transformed) into a format that can be analyzed, and stored (loaded) into a data warehouse or other system. Extract, load, transform (ELT) is an alternate but related approach designed to push processing down to the database for improved performance.

#### Project Proposal
- Combine the Best 50 Superhero Movies dataset with The Superhero dataset for the analysis and data visualization.

#### 3 questions relating to the data:

For the 3 questions, I decided to focus exclusively on American superheroes and movies.

1. Is there a gender bias in superhero movie industry?

2. What are the most important variables in predicting movie's profitability.

3. For production companies that were acquired, how much on average has each superhero earned?



#### Extract Data:
Data sources:
•	Superheroes API 
•	Web Scraping
•	Movies API 
•	Python package - IMDbPY 

Data formats:
•	CSV
•	XLSX
•	pgAdmin
•	HTML
        
 
#### Transform Data: 
•	cleaning
•	joining 
•	filtering
•	aggregating (group by, sort, average, count)
•	deleting columns
•	data visualization 


#### Load Data:
           load the final database into the PostgreSQL database server.


#### Flask Application:
           render images for each movie, superhero and a short movie summary from the database.


In [4]:
# import dependencies 
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import requests
import time
from scipy.stats import linregress
from pprint import pprint
import json

## Data Preparation (general)

In [16]:
# PART 1 =====================================
# First sorce of data comes from superhero API.
# requests library helps us get the content from the API by using the get() method. 

url = "https://superhero-search.p.rapidapi.com/"

querystring = {"hero":"spiderman"}



response = requests.request("GET", url, headers=headers, params=querystring)

print(response.text)

{"id":476,"name":"Spider-Man","slug":"620-spider-man","powerstats":{"intelligence":90,"strength":55,"speed":67,"durability":75,"power":74,"combat":85},"appearance":{"gender":"Male","race":"Human","height":["5'10","178 cm"],"weight":["165 lb","74 kg"],"eyeColor":"Hazel","hairColor":"Brown"},"biography":{"fullName":"Peter Parker","alterEgos":"No alter egos found.","aliases":["Spiderman","Bag-Man","Black Marvel","Captain Universe","Dusk","Green Hood","Hornet","Mad Dog 336","Peter Palmer","Prodigy","Ricochet","Scarlet Spider","Spider-Boy","Spider-Hulk","Spider-Morphosis"],"placeOfBirth":"New York, New York","firstAppearance":"Amazing Fantasy #15","publisher":"Marvel Comics","alignment":"good"},"work":{"occupation":"Freelance photographer, teacher","base":"New York, New York"},"connections":{"groupAffiliation":"Member of the Avengers, formerly member of Outlaws, alternate Fantastic Four","relatives":"Richard Parker (father, deceased), Mary Parker(mother, deceased), Benjamin Parker (uncle, d

In [17]:
# Python has a built-in package called json, which can be used to work with JSON data we get from the API.
response = requests.get(url, headers=headers, params=querystring).json()
print(json.dumps(response, indent=4, sort_keys=True))

{
    "appearance": {
        "eyeColor": "Hazel",
        "gender": "Male",
        "hairColor": "Brown",
        "height": [
            "5'10",
            "178 cm"
        ],
        "race": "Human",
        "weight": [
            "165 lb",
            "74 kg"
        ]
    },
    "biography": {
        "aliases": [
            "Spiderman",
            "Bag-Man",
            "Black Marvel",
            "Captain Universe",
            "Dusk",
            "Green Hood",
            "Hornet",
            "Mad Dog 336",
            "Peter Palmer",
            "Prodigy",
            "Ricochet",
            "Scarlet Spider",
            "Spider-Boy",
            "Spider-Hulk",
            "Spider-Morphosis"
        ],
        "alignment": "good",
        "alterEgos": "No alter egos found.",
        "firstAppearance": "Amazing Fantasy #15",
        "fullName": "Peter Parker",
        "placeOfBirth": "New York, New York",
        "publisher": "Marvel Comics"
    },
    "connections": {
   

In [18]:
# set up lists to hold reponse info
hero_id = []
name = []
fullName = []
gender = []
eyeColor = []
hairColor = []
height = []
weight = []
race = []
placeOfBirth = []
work = []
images = []
publisher = []
intelligence = []
power = []
firstAppearance = []
count = 0
count_error = 0



# Loop through the list of 561 heroes and perform a request for each hero
for i in range(561):
    
        url = "https://superhero-search.p.rapidapi.com/"

        querystring = {"id": i}

        headers = {
                    'x-rapidapi-host': "superhero-search.p.rapidapi.com",
                    'x-rapidapi-key': "5a8ea71761msh2e12845540e03c1p1bacdfjsn766f6824866c"
                    }

        response = requests.get(url, headers=headers, params=querystring).json()
#         print(json.dumps(response, indent=4, sort_keys=True))
        try:
    #         pprint(response)
#             print('==============')



            id_v = (response['id'])
            name_v = (response['name'])
            fullName_v = (response['biography']['fullName'])
            gender_v = response['appearance']['gender']
            eyeColor_v = response['appearance']['eyeColor']
            hairColor_v =  response['appearance']['hairColor']
            height_v = response['appearance']['height'][0]
            weight_v = response['appearance']['weight'][0]
            race_v = response['appearance']['race']
            placeOfBirth_v = (response['biography']['placeOfBirth'])
            work_v = response['work']['occupation']
            images_v =  response['images']['lg']
            publisher_v = (response['biography']['publisher'])
            firstAppearance_v = response['biography']['firstAppearance'] 
            intelligence_v = response['powerstats']['intelligence']
            power_v = response['powerstats']['power']
            

            hero_id.append(id_v)
            name.append(name_v)
            fullName.append(fullName_v)
            gender.append(gender_v)
            eyeColor.append(eyeColor_v)
            hairColor.append(fullName_v)
            height.append(height_v)
            weight.append(weight_v)
            race.append(race_v)
            placeOfBirth.append(placeOfBirth_v)
            work.append(work_v)
            images.append(images_v)
            publisher.append(publisher_v)
            intelligence.append(intelligence_v)
            power.append(power_v)
            firstAppearance.append(firstAppearance_v)
#             print(count)
            count +=1


        except (KeyError, IndexError):
            count_error +=1
            print(f"Missing field/result... skipping. {count_error}") 

print(f"Total results :{count} heros")        
print(f"Missing field/result:{count_error} errors") 
        

Total results :561 heros
Missing field/result:0 errors


In [35]:
# creating a data frame for heroes
hero_dict = {
        "hero_id" : hero_id,
        "name" : name,
        "fullName":fullName,
        "gender":gender,
        "eyeColor" :eyeColor,
        "hairColor":hairColor,
        "height":height,
        "weight":weight,
        "race" :race,
        "placeOfBirth":placeOfBirth,
        "work" : work,
        "images": images,
        "publisher": publisher,
        "intelligence":intelligence,
        "power":power,
        "firstAppearance":firstAppearance
}

hero_df = pd.DataFrame(hero_dict)
print(hero_df.count())
hero_df.head()

hero_id            561
name               561
fullName           561
gender             561
eyeColor           561
hairColor          561
height             561
weight             561
race               398
placeOfBirth       561
work               561
images             561
publisher          555
intelligence       561
power              561
firstAppearance    561
dtype: int64


hero_id         name                fullName gender eyeColor  \
0        0       A-Bomb  Richard Milhouse Jones   Male   Yellow   
1        1   Abe Sapien          Abraham Sapien   Male     Blue   
2        2     Abin Sur                           Male     Blue   
3        3  Abomination            Emil Blonsky   Male    Green   
4        4      Abraxas                 Abraxas   Male     Blue   

                hairColor height  weight               race  \
0  Richard Milhouse Jones    6'8  980 lb              Human   
1          Abraham Sapien    6'3  145 lb      Icthyo Sapien   
2                            6'1  200 lb            Ungaran   
3            Emil Blonsky    6'8  980 lb  Human / Radiation   
4                 Abraxas      -    - lb      Cosmic Entity   

         placeOfBirth                                               work  \
0  Scarsdale, Arizona  Musician, adventurer, author; formerly talk sh...   
1                   -                            Paranormal Investigator   
2              Ungara            Green Lantern, former history professor   
3  Zagreb, Yugoslavia                                             Ex-Spy   
4     Within Eternity                              Dimensional destroyer   

                                              images          publisher  \
0  https://cdn.rawgit.com/akabab/superhero-api/0....      Marvel Comics   
1  https://cdn.rawgit.com/akabab/superhero-api/0....  Dark Horse Comics   
2  https://cdn.rawgit.com/akabab/superhero-api/0....          DC Comics   
3  https://cdn.rawgit.com/akabab/superhero-api/0....      Marvel Comics   
4  https://cdn.rawgit.com/akabab/superhero-api/0....      Marvel Comics   

   intelligence  power                          firstAppearance  
0            38     24  Hulk Vol 2 #2 (April, 2008) (as A-Bomb)  
1            88    100      Hellboy: Seed of Destruction (1993)  
2            50     99             Showcase #22 (October, 1959)  
3            63     62                    Tales to Astonish #90  
4            88    100              Fantastic Four Annual #2001

## Data Preparation (Question 1)

In [36]:
# Data cleaning for hero_df. Deleting some columns, that do not have valuable info
hero_df = hero_df.drop(['hero_id', 'hairColor', 'firstAppearance', 'publisher'], axis=1)
hero_df.head()


name                fullName gender eyeColor height  weight  \
0       A-Bomb  Richard Milhouse Jones   Male   Yellow    6'8  980 lb   
1   Abe Sapien          Abraham Sapien   Male     Blue    6'3  145 lb   
2     Abin Sur                           Male     Blue    6'1  200 lb   
3  Abomination            Emil Blonsky   Male    Green    6'8  980 lb   
4      Abraxas                 Abraxas   Male     Blue      -    - lb   

                race        placeOfBirth  \
0              Human  Scarsdale, Arizona   
1      Icthyo Sapien                   -   
2            Ungaran              Ungara   
3  Human / Radiation  Zagreb, Yugoslavia   
4      Cosmic Entity     Within Eternity   

                                                work  \
0  Musician, adventurer, author; formerly talk sh...   
1                            Paranormal Investigator   
2            Green Lantern, former history professor   
3                                             Ex-Spy   
4                              Dimensional destroyer   

                                              images  intelligence  power  
0  https://cdn.rawgit.com/akabab/superhero-api/0....            38     24  
1  https://cdn.rawgit.com/akabab/superhero-api/0....            88    100  
2  https://cdn.rawgit.com/akabab/superhero-api/0....            50     99  
3  https://cdn.rawgit.com/akabab/superhero-api/0....            63     62  
4  https://cdn.rawgit.com/akabab/superhero-api/0....            88    100

In [37]:
# Saving the dataframe to heroes.csv file
hero_df.to_csv('heroes.csv', index=False) 

In [38]:
# Creating a test data frame from heroes.csv 
test_df = pd.read_csv('heroes.csv')

# Data Cleanup & Analysis
test_df.loc[test_df['gender'] == '-', 'gender'] = 'Other'
test_df.head()


name                fullName gender eyeColor height  weight  \
0       A-Bomb  Richard Milhouse Jones   Male   Yellow    6'8  980 lb   
1   Abe Sapien          Abraham Sapien   Male     Blue    6'3  145 lb   
2     Abin Sur                     NaN   Male     Blue    6'1  200 lb   
3  Abomination            Emil Blonsky   Male    Green    6'8  980 lb   
4      Abraxas                 Abraxas   Male     Blue      -    - lb   

                race        placeOfBirth  \
0              Human  Scarsdale, Arizona   
1      Icthyo Sapien                   -   
2            Ungaran              Ungara   
3  Human / Radiation  Zagreb, Yugoslavia   
4      Cosmic Entity     Within Eternity   

                                                work  \
0  Musician, adventurer, author; formerly talk sh...   
1                            Paranormal Investigator   
2            Green Lantern, former history professor   
3                                             Ex-Spy   
4                              Dimensional destroyer   

                                              images  intelligence  power  
0  https://cdn.rawgit.com/akabab/superhero-api/0....            38     24  
1  https://cdn.rawgit.com/akabab/superhero-api/0....            88    100  
2  https://cdn.rawgit.com/akabab/superhero-api/0....            50     99  
3  https://cdn.rawgit.com/akabab/superhero-api/0....            63     62  
4  https://cdn.rawgit.com/akabab/superhero-api/0....            88    100

In [23]:
# Groupby gender and calculating avarage intelligence for each group
test_df.groupby('gender')['intelligence'].mean()
test_df.groupby('gender')['name'].count()
gender_df = pd.DataFrame()
gender_df['count'] = test_df.groupby('gender')['name'].count()
gender_df['avr intelligence'] = round(test_df.groupby('gender')['intelligence'].mean(), 2)
gender_df =gender_df.reset_index()
gender_df

gender  count  avr intelligence
0  Female    151             62.50
1    Male    395             65.42
2   Other     15             60.40

In [33]:
import plotly.graph_objects as go
x = ['Female', 'Male', 'Other']


fig = go.Figure(data=[
    go.Bar(name='count', x=x, y = gender_df['count']),
    go.Bar(name='intelligence', x=x, y = gender_df['avr intelligence'])
])
# Change the bar mode
fig.update_layout(barmode='group')


fig.update_layout(
    font=dict(
        family="Courier New, monospace",
        size=18,
        color="#7f7f7f"
    ),
    title={
        'text':'SuperHeroes Counts and Intelligence by Gender',
        'y':0.9,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'}
)

fig.update_xaxes(title_text='Gender')
fig.update_yaxes(title_text='Count / Intelligence')

fig.show()

In [39]:
import plotly.express as px
df = test_df

fig = px.scatter(df, x="power", y="weight", facet_col="gender")
fig.update_xaxes(title_font=dict(size=18, family='Courier', color='crimson'))
fig.update_yaxes(title_font=dict(size=18, family='Courier', color='red'))


fig.update_layout(
    font=dict(
        family="Courier New, monospace",
        size=18,
        color="#7f7f7f"
    ),
    title={
        'text':'SuperHeroes Power vs Weight by Gender',
        'y':1,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'}
)

fig.show()


 #### Data Analysis
1. I found that there is gender inequality in the Superheroes dataset. There are 395 male heroes, 151 female heroes and 15 other gender heroes. The average intelligence for a male hero is 65.42 points. It is higher than the average intelligence for a female   hero (62.50). Clearly, there are the immediate issues of parity, equal treatment and human rights.  
 
2. Looks like there is no strong correlation between the power of hero and his/her/other weight. 
 
3. There is no question that watching films can help us escape our everyday lives. However, all films impact our society and popular culture. I am sure that a growing number of female superheroes with a high intelligence will empower girls and women all around the world.

In [41]:
# PART 2 =====================================
# Second sorce of data comes from scraping a web page with a list of best 50 Superheroes movies af all time.

from bs4 import BeautifulSoup
import requests

url = 'http://superheroes.theringer.com/?_ga=2.205407573.49282893.1583382995-1394544322.1583382995'

In [42]:
# Retrieve page with the requests module
response = requests.get(url)

In [43]:
# Create BeautifulSoup object; 
# parse with 'html.parser'
soup = BeautifulSoup(response.text, 'html.parser')

In [47]:
# Examine the results
# print(soup.prettify())

In [48]:
# results are returned as an iterable list
results = soup.find_all('li', class_='card-item')

# results[0].text
# results[0]

In [49]:
# Creating lists to hold the info
rank = []
title= []
date= []
rottenTomatoes= []
boxOffice= []
rewatchability= []
totalScore= []
image= []

# Loop through returned results
for result in results:
#     print(result)
    # Error handling
    try:
        # Identify and return title of the movie
        title_v = result.find('span', class_="title").text
#         print(title_v)
        # Identify and return date of the movie
        date_v = result.find('span', class_="date").text
#         print(date_v)
        
        rank_v = result.find('div', class_="rank").span.text
#         print(rank_v)
        
        rottenTomatoes_v = result.find('span', class_="data").text
#         print(rottenTomatoes_v)
    
        boxOfficeB = result.find('div', class_='expanded-data-col adjusted')
        boxOffice_v = boxOfficeB.find('span', class_='data offset').text
#         print(boxOffice_v)
        
        rewatchabilityB = result.find('div', class_='expanded-data-col rewatchability')
        rewatchability_v = rewatchabilityB.find('span', class_='data').text
#         print(rewatchability_v)
        
        totalScoreB = result.find('div', class_='expanded-data-col score')
        totalScore_v = totalScoreB.find('span', class_='data').text
#         print(totalScore_v)
        
        # Identify and return an image link 
        image_v = result.find('div', class_="image")['data-src']
#         print(image_v)
        
        

        rank.append(rank_v)
        title.append(title_v)
        date.append(date_v)
        rottenTomatoes.append(rottenTomatoes_v)
        boxOffice.append(boxOffice_v)
        rewatchability.append(rewatchability_v)
        totalScore.append(totalScore_v)
        image.append(image_v)

        
        
        # Print results only if items are available
#         if (title and date and rank and rottenTomatoes and boxOffice and rewatchability and totalScore and image):
#             print('-------------')
#             print(rank)
#             print(title)
#             print(date)
#             print(rottenTomatoes)
#             print(boxOffice)
#             print(rewatchability)
#             print(totalScore)
#             print(image)
    except AttributeError as e:
        print(e)

In [53]:
# create a data frame for the best 50 movies
best_50_movies_dict = {
        "rank" : rank,
        "title" : title,
        "date":date,
        "rottenTomatoes":rottenTomatoes,
        "boxOffice" :boxOffice,
        "rewatchability":rewatchability,
        "totalScore":totalScore,
        "image":image
}

movies_df = pd.DataFrame(best_50_movies_dict)
movies_df.to_csv('movies.csv', index=False)
print(movies_df.count())
movies_df.head(50)


rank              50
title             50
date              50
rottenTomatoes    50
boxOffice         50
rewatchability    50
totalScore        50
image             50
dtype: int64


rank                                title  date rottenTomatoes boxOffice  \
0    50                             The Crow  1994            82%    $83.3M   
1    49                       Batman Forever  1995            41%     $295M   
2    48          Hellboy II: The Golden Army  2008            85%      $87M   
3    47                 Thor: The Dark World  2013            66%     $219M   
4    46                         Man of Steel  2013            55%     $309M   
5    45                                 Thor  2011            77%     $199M   
6    44                              Shazam!  2019            91%   $140.4M   
7    43                X-Men: The Last Stand  2006            58%     $286M   
8    42                              Ant-Man  2015            82%     $183M   
9    41               The Amazing Spider-Man  2012            72%     $282M   
10   40   Captain America: The First Avenger  2011            80%     $194M   
11   39                     Superman Returns  2006            76%     $286M   
12   38                   X-Men: First Class  2011            86%     $161M   
13   37                                X-Men  2000            81%     $223M   
14   36                 Ant-Man and the Wasp  2018            88%     $217M   
15   35                              Aquaman  2018            64%     $335M   
16   34                         Spider-Man 3  2007            63%     $399M   
17   33                       Doctor Strange  2016            89%     $233M   
18   32                       Batman Returns  1992            81%     $283M   
19   31                        Batman Begins  2005            84%     $256M   
20   30           X-Men: Days of Future Past  2014            90%     $244M   
21   29    Spider-Man: Into the Spider-Verse  2018            97%     $190M   
22   28                          Superman II  1981            87%     $290M   
23   27                     X2: X-Men United  2003            86%     $287M   
24   26                           Iron Man 2  2010            73%     $353M   
25   25                           Deadpool 2  2018            83%     $325M   
26   24                                Logan  2017            93%     $226M   
27   23                       Captain Marvel  2019            78%     $427M   
28   22  Captain America: The Winter Soldier  2014            89%     $271M   
29   21              Avengers: Age of Ultron  2015            75%     $467M   
30   20                       Thor: Ragnarok  2017            93%     $315M   
31   19       Guardians of the Galaxy Vol. 2  2017            82%     $389M   
32   18                               Batman  1989            72%     $493M   
33   17                             Deadpool  2016            84%     $363M   
34   16                           Iron Man 3  2013            79%     $434M   
35   15            Spider-Man: Far From Home  2019            90%     $373M   
36   14               Spider-Man: Homecoming  2017            92%     $333M   
37   13              Guardians of the Galaxy  2014            91%     $348M   
38   12                The Dark Knight Rises  2012            87%     $482M   
39   11           Captain America: Civil War  2016            90%     $408M   
40   10                             Iron Man  2008            94%     $364M   
41    9               Avengers: Infinity War  2018            84%     $679M   
42    8                         Wonder Woman  2017            92%     $412M   
43    7                         Spider-Man 2  2004            94%     $483M   
44    6                           Spider-Man  2002            89%     $550M   
45    5                             Superman  1978            93%     $502M   
46    4                         The Avengers  2012            92%     $671M   
47    3                    Avengers: Endgame  2019            94%     $858M   
48    2                        Black Panther  2018            96%     $700M   
49    1                      The Dark Knight  2008   

## Data Preparation (Question 2)

In [54]:
# Data Cleanup & Analysis
# Getting a value for the boxOffice column. Cutting '$' and 'M' : $83.3M  will make 83.3
movies_df['boxOffice'] = movies_df['boxOffice'].str[1:-1].astype(float)
movies_df.head()


rank                        title  date rottenTomatoes  boxOffice  \
0   50                     The Crow  1994            82%       83.3   
1   49               Batman Forever  1995            41%      295.0   
2   48  Hellboy II: The Golden Army  2008            85%       87.0   
3   47         Thor: The Dark World  2013            66%      219.0   
4   46                 Man of Steel  2013            55%      309.0   

  rewatchability totalScore                                              image  
0           6.25        125  https://hardrefreshcdn.com/public/media/superh...  
1           5.82        130  https://hardrefreshcdn.com/public/media/superh...  
2           7.00        134  https://hardrefreshcdn.com/public/media/superh...  
3           4.45        134  https://hardrefreshcdn.com/public/media/superh...  
4           3.91        139  https://hardrefreshcdn.com/public/media/superh...

In [57]:
# This chart contains the top movies based on the cumulative worldwide box office vs  rotten tomatoes ratings,  and total score
import plotly.express as px
from plotly.subplots import make_subplots
x =movies_df['title']

# Create traces
fig = go.Figure()

# Create figure with secondary y-axis
fig = make_subplots(specs=[[{"secondary_y": True}]])


fig.add_trace(go.Scatter(x=x, y=movies_df['boxOffice'] + 5,
                    mode='lines',
                    name='Box Office'),
                    secondary_y=False,)
fig.add_trace(go.Scatter(x=x, y=movies_df['rottenTomatoes'],
                    mode='lines+markers',
                    name='Rotten Tomatoes'),
                    secondary_y=True,)
fig.add_trace(go.Scatter(x=x, y=movies_df['totalScore'],
                    mode='lines+markers',
                    name='Total Score'),
                    secondary_y=True,)
# fig.add_trace(go.Scatter(x=x, y=movies_df['rewatchability'],
#                     mode='lines+markers',
#                     name='Rewatchability'),
#                     secondary_y=True,)


fig.update_layout(
    font=dict(
        family="Courier New, monospace",
        size=18,
        color="#7f7f7f"
    ),
    title={
        'text':'Box office vs Movie Rating',
        'y':1,
        'x':0.4,
        'xanchor': 'center',
        'yanchor': 'top'})

# fig.update_xaxes(title_text='Movies Titles')

# Set y-axes titles
fig.update_yaxes(title_text="Box Office $M", secondary_y=False)
fig.update_yaxes(title_text="Scores", secondary_y=True)

# fig.update_yaxes(title_text='Scores')
fig.update_layout(
    margin=dict(l=20, r=20, t=20, b=200)
    
)


fig.show()

In [63]:
# Getting a most profitable year
box = movies_df.groupby('date')['boxOffice'].sum()
box_df = pd.DataFrame(box)
box_df =box_df.reset_index()
box_df = box_df.sort_values(by=['boxOffice'], ascending=False)
box_df.head()

date  boxOffice
22  2018     2446.0
23  2019     1798.4
21  2017     1675.0
16  2012     1435.0
13  2008     1063.0

In [64]:
# Getting a most profitable movie
movies_df = movies_df.sort_values(by=['boxOffice'], ascending=False)
movies_df.head()

rank                   title  date rottenTomatoes  boxOffice  \
47    3       Avengers: Endgame  2019            94%      858.0   
48    2           Black Panther  2018            96%      700.0   
41    9  Avengers: Infinity War  2018            84%      679.0   
46    4            The Avengers  2012            92%      671.0   
49    1         The Dark Knight  2008            94%      612.0   

   rewatchability totalScore  \
47           7.50        212   
48           9.60        213   
41            8.0        198   
46           7.36        210   
49           9.27        214   

                                                image  
47  https://hardrefreshcdn.com/public/media/superh...  
48  https://hardrefreshcdn.com/public/media/superh...  
41  https://hardrefreshcdn.com/public/media/superh...  
46  https://hardrefreshcdn.com/public/media/superh...  
49  https://hardrefreshcdn.com/public/media/superh...

In [65]:
# Compare box office over the years
import plotly.express as px
df = movies_df
fig = px.scatter(df, x="date", y="boxOffice", size="boxOffice", color="boxOffice",
           hover_name="title", log_x=True, size_max=60)


fig.update_xaxes(title_text='year')
fig.update_yaxes(title_text='Box Office ($million)')

fig.update_layout(
    title={
        'text':'Box Office Over Time',
        'y':0.95,
        'x':0.4,
        'xanchor': 'center',
        'yanchor': 'top'}
)


fig.show()

### Data Analysis
1. The rotten tomatoes score didn't make a strong move with the time, while the total score gets higher with each year.
2. Box office performance gets bigger with each year. The movie - Avengers: Endgame (2019) brought $858 M.
3. For the last 15 years we can see an increasing production of superhero movies with much bigger box office every year. Looks like a money-making-machine for movie investors, which brings us to the conclusion - we are going to see even more superhero movies in the future.


In [70]:
# PART 3 =====================================
# Third sorce of data comes from omdb API for more information about our 50 movies. 
# In this section I am also using IMDbPY.
# IMDbPY is a Python package for retrieving and managing the data about movies and people.

# Note that the ?t= is a query param for the t-itle of the
# movie we want to search for.
url = "http://www.omdbapi.com/?t="
api_key = "&apikey=trilogy"

# Example of response for movie 'The Matrix'
response = requests.get(url + "The Matrix" + api_key)
print(response.url)
data = response.json()
pprint(data)

http://www.omdbapi.com/?t=The%20Matrix&apikey=trilogy
{'Actors': 'Keanu Reeves, Laurence Fishburne, Carrie-Anne Moss, Hugo Weaving',
 'Awards': 'Won 4 Oscars. Another 37 wins & 50 nominations.',
 'BoxOffice': 'N/A',
 'Country': 'USA',
 'DVD': '21 Sep 1999',
 'Director': 'Lana Wachowski, Lilly Wachowski',
 'Genre': 'Action, Sci-Fi',
 'Language': 'English',
 'Metascore': '73',
 'Plot': 'A computer hacker learns from mysterious rebels about the true '
         'nature of his reality and his role in the war against its '
         'controllers.',
 'Poster': 'https://m.media-amazon.com/images/M/MV5BNzQzOTk3OTAtNDQ0Zi00ZTVkLWI0MTEtMDllZjNkYzNjNTc4L2ltYWdlXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SX300.jpg',
 'Production': 'Warner Bros. Pictures',
 'Rated': 'R',
 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '8.7/10'},
             {'Source': 'Rotten Tomatoes', 'Value': '88%'},
             {'Source': 'Metacritic', 'Value': '73/100'}],
 'Released': '31 Mar 1999',
 'Response': 'True',
 'Run

In [73]:
import imdb
movies_titles = title
print(title)

# Creating a lists to store the data
imdbID=[]
id_clean = []
Title=[]
Year=[]
Production=[]
Plot=[]
Poster=[]
Actor = []
Role1 = []
Role2 = []
Actors = []
count = 0


['The Crow', 'Batman Forever', 'Hellboy II: The Golden Army', 'Thor: The Dark World', 'Man of Steel', 'Thor', 'Shazam!', 'X-Men: The Last Stand', 'Ant-Man', 'The Amazing Spider-Man', 'Captain America: The First Avenger', 'Superman Returns', 'X-Men: First Class', 'X-Men', 'Ant-Man and the Wasp', 'Aquaman', 'Spider-Man 3', 'Doctor Strange', 'Batman Returns', 'Batman Begins', 'X-Men: Days of Future Past', 'Spider-Man: Into the Spider-Verse', 'Superman II', 'X2: X-Men United', 'Iron Man 2', 'Deadpool 2', 'Logan', 'Captain Marvel', 'Captain America: The Winter Soldier', 'Avengers: Age of Ultron', 'Thor: Ragnarok', 'Guardians of the Galaxy Vol. 2', 'Batman', 'Deadpool', 'Iron Man 3', 'Spider-Man: Far From Home', 'Spider-Man: Homecoming', 'Guardians of the Galaxy', 'The Dark Knight Rises', 'Captain America: Civil War', 'Iron Man', 'Avengers: Infinity War', 'Wonder Woman', 'Spider-Man 2', 'Spider-Man', 'Superman', 'The Avengers', 'Avengers: Endgame', 'Black Panther', 'The Dark Knight']


In [74]:
# Performing a GET request similar to the one we executed
for i in movies_titles:
    # Error handling
        try:
            response = requests.get(url + i + api_key)
#             print(response.url)
            # Converting the response to JSON, and printing the result.
            data = response.json()
            print(data)
            imdbID_v = data['imdbID']
            id_v = imdbID_v[2:]
        
            
            Title_v = data['Title']
            Year_v = data['Year']
            Production_v = data['Production']
            Plot_v = data['Plot']
            Poster_v = data['Poster']
            Actors_v = data['Actors']
            
#             Getting actor and current role
            i =  imdb.IMDb(accessSystem='http')
            movie = i.get_movie(id_v)
            m_title = movie['title']
            print("=============")
            print(count)
            print(Poster_v)
            
            # Get the 1st Person object in the cast list
            cast = movie['cast'][0]
            cast
#             print(cast['name'])
            Actor_v = cast['name']
#             print(cast.currentRole)
            
            
            if len(cast.currentRole) == 1:
                Role1_v = str(cast.currentRole)
                Role2_v = 'None'
            elif len(cast.currentRole) == 2:
                Role1_v = str(cast.currentRole[0])
                Role2_v = str(cast.currentRole[1])
            else:
                Role1_v = 'None'
                Role2_v = 'None'
                
              # Get the 2nd Person object in the cast list
            second_cast = movie['cast'][1]
            second_cast
#             print(second_cast['name'])
#             Actor_v = second_cast['name']
                      

#             print(f"Movie title: {m_title}")
            print(f"Movie title: {Title_v}")
            print(m_title)
            print(id_v)
#             print(imdbID_v)
#             print(movie['cast'])
            print(Actors_v)
    
    
    
            imdbID.append(imdbID_v)
            id_clean.append(id_v)
            Title.append(Title_v)
            Year.append(Year_v)
            Actor.append(Actor_v)
            
            Role1.append(Role1_v)
            Role2.append(Role2_v)
#             print(cast.currentRole)
            print(f"{Actor_v} : {Role1_v} : {Role2_v}")
            print(f"Second_cast actor: {second_cast['name']} : {second_cast.currentRole}")
            print("===end=====")
            Production.append(Production_v)
            Plot.append(Plot_v)
            Poster.append(Poster_v)
            count +=1
        
        except AttributeError as e:
            print(e)


{'Title': 'The Crow', 'Year': '1994', 'Rated': 'R', 'Released': '13 May 1994', 'Runtime': '102 min', 'Genre': 'Action, Drama, Fantasy', 'Director': 'Alex Proyas', 'Writer': "James O'Barr (comic book series and comic strip), David J. Schow (screenplay), John Shirley (screenplay)", 'Actors': 'Brandon Lee, Rochelle Davis, Ernie Hudson, Michael Wincott', 'Plot': "A man brutally murdered comes back to life as an undead avenger of his and his fiancée's murder.", 'Language': 'English', 'Country': 'USA', 'Awards': '5 wins & 7 nominations.', 'Poster': 'https://m.media-amazon.com/images/M/MV5BM2Y4ZGVhZjItNjU0OC00MDk1LWI4ZTktYTgwMWJkNDE5OTcxXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg', 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '7.6/10'}, {'Source': 'Rotten Tomatoes', 'Value': '82%'}, {'Source': 'Metacritic', 'Value': '71/100'}], 'Metascore': '71', 'imdbRating': '7.6', 'imdbVotes': '162,523', 'imdbID': 'tt0109506', 'Type': 'movie', 'DVD': '01 Jan 1998', 'BoxOffice': 'N/A', 'Product

5
https://m.media-amazon.com/images/M/MV5BOGE4NzU1YTAtNzA3Mi00ZTA2LTg2YmYtMDJmMThiMjlkYjg2XkEyXkFqcGdeQXVyNTgzMDMzMTg@._V1_SX300.jpg
Movie title: Thor
Thor
0800369
Chris Hemsworth, Natalie Portman, Tom Hiddleston, Anthony Hopkins
Chris Hemsworth : Thor : None
Second_cast actor: Natalie Portman : Jane Foster
===end=====
{'Title': 'Shazam!', 'Year': '2019', 'Rated': 'PG-13', 'Released': '05 Apr 2019', 'Runtime': '132 min', 'Genre': 'Action, Adventure, Comedy, Fantasy', 'Director': 'David F. Sandberg', 'Writer': 'Henry Gayden (screenplay by), Henry Gayden (story by), Darren Lemke (story by), Bill Parker (Shazam created by), C.C. Beck (Shazam created by)', 'Actors': 'Zachary Levi, Mark Strong, Asher Angel, Jack Dylan Grazer', 'Plot': 'A newly fostered young boy in search of his mother instead finds unexpected super powers and soon gains a powerful enemy.', 'Language': 'English, Spanish', 'Country': 'USA, Canada', 'Awards': '21 nominations.', 'Poster': 'https://m.media-amazon.com/images/M/M

11
https://m.media-amazon.com/images/M/MV5BNzY2ZDQ2MTctYzlhOC00MWJhLTgxMmItMDgzNDQwMDdhOWI2XkEyXkFqcGdeQXVyNjc1NTYyMjg@._V1_SX300.jpg
Movie title: Superman Returns
Superman Returns
0348150
Brandon Routh, Kate Bosworth, Kevin Spacey, James Marsden
Brandon Routh : Clark Kent : Superman
Second_cast actor: Kate Bosworth : Lois Lane
===end=====
{'Title': 'X-Men: First Class', 'Year': '2011', 'Rated': 'PG-13', 'Released': '03 Jun 2011', 'Runtime': '131 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Matthew Vaughn', 'Writer': 'Ashley Miller (screenplay by), Zack Stentz (screenplay by), Jane Goldman (screenplay by), Matthew Vaughn (screenplay by), Sheldon Turner (story by), Bryan Singer (story by)', 'Actors': 'James McAvoy, Laurence Belcher, Michael Fassbender, Bill Milner', 'Plot': "In the 1960s, superpowered humans Charles Xavier and Erik Lensherr work together to find others like them, but Erik's vengeful pursuit of an ambitious mutant who ruined his life causes a schism to divide

17
https://m.media-amazon.com/images/M/MV5BNjgwNzAzNjk1Nl5BMl5BanBnXkFtZTgwMzQ2NjI1OTE@._V1_SX300.jpg
Movie title: Doctor Strange
Doctor Strange
1211837
Benedict Cumberbatch, Chiwetel Ejiofor, Rachel McAdams, Benedict Wong
Benedict Cumberbatch : Dr. Stephen Strange : None
Second_cast actor: Chiwetel Ejiofor : Mordo
===end=====
{'Title': 'Batman Returns', 'Year': '1992', 'Rated': 'PG-13', 'Released': '19 Jun 1992', 'Runtime': '126 min', 'Genre': 'Action, Crime, Fantasy', 'Director': 'Tim Burton', 'Writer': 'Bob Kane (Batman characters), Daniel Waters (story), Sam Hamm (story), Daniel Waters (screenplay)', 'Actors': 'Michael Keaton, Danny DeVito, Michelle Pfeiffer, Christopher Walken', 'Plot': 'Batman returns to the big screen when a deformed man calling himself the Penguin wreaks havoc across Gotham with the help of a cruel businessman.', 'Language': 'English', 'Country': 'USA, UK', 'Awards': 'Nominated for 2 Oscars. Another 2 wins & 27 nominations.', 'Poster': 'https://m.media-amazon.c

23
https://m.media-amazon.com/images/M/MV5BNDk0NjYxMzIzOF5BMl5BanBnXkFtZTYwMTc1MjU3._V1_SX300.jpg
Movie title: X2: X-Men United
X2: X-Men United
0290334
Patrick Stewart, Hugh Jackman, Ian McKellen, Halle Berry
Patrick Stewart : Professor Charles Xavier : None
Second_cast actor: Hugh Jackman : Logan / Wolverine
===end=====
{'Title': 'Iron Man 2', 'Year': '2010', 'Rated': 'PG-13', 'Released': '07 May 2010', 'Runtime': '124 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Jon Favreau', 'Writer': 'Justin Theroux (screenplay), Stan Lee (Marvel comic book), Don Heck (Marvel comic book), Larry Lieber (Marvel comic book), Jack Kirby (Marvel comic book)', 'Actors': 'Robert Downey Jr., Gwyneth Paltrow, Don Cheadle, Scarlett Johansson', 'Plot': "With the world now aware of his identity as Iron Man, Tony Stark must contend with both his declining health and a vengeful mad man with ties to his father's legacy.", 'Language': 'English, French, Russian', 'Country': 'USA', 'Awards': 'Nominated 

29
https://m.media-amazon.com/images/M/MV5BMTM4OGJmNWMtOTM4Ni00NTE3LTg3MDItZmQxYjc4N2JhNmUxXkEyXkFqcGdeQXVyNTgzMDMzMTg@._V1_SX300.jpg
Movie title: Avengers: Age of Ultron
Avengers: Age of Ultron
2395427
Robert Downey Jr., Chris Hemsworth, Mark Ruffalo, Chris Evans
Robert Downey Jr. : Tony Stark : Iron Man
Second_cast actor: Chris Hemsworth : Thor
===end=====
{'Title': 'Thor: Ragnarok', 'Year': '2017', 'Rated': 'PG-13', 'Released': '03 Nov 2017', 'Runtime': '130 min', 'Genre': 'Action, Adventure, Comedy, Fantasy, Sci-Fi', 'Director': 'Taika Waititi', 'Writer': 'Eric Pearson, Craig Kyle, Christopher L. Yost, Stan Lee (based on the Marvel comics by), Larry Lieber (based on the Marvel comics by), Jack Kirby (based on the Marvel comics by)', 'Actors': 'Chris Hemsworth, Tom Hiddleston, Cate Blanchett, Idris Elba', 'Plot': 'Imprisoned on the planet Sakaar, Thor must race against time to return to Asgard and stop Ragnarök, the destruction of his world, at the hands of the powerful and ruthless

35
https://m.media-amazon.com/images/M/MV5BMGZlNTY1ZWUtYTMzNC00ZjUyLWE0MjQtMTMxN2E3ODYxMWVmXkEyXkFqcGdeQXVyMDM2NDM2MQ@@._V1_SX300.jpg
Movie title: Spider-Man: Far from Home
Spider-Man: Far from Home
6320628
Tom Holland, Samuel L. Jackson, Jake Gyllenhaal, Marisa Tomei
Tom Holland : Peter Parker : Spider-Man
Second_cast actor: Samuel L. Jackson : Nick Fury
===end=====
{'Title': 'Spider-Man: Homecoming', 'Year': '2017', 'Rated': 'PG-13', 'Released': '07 Jul 2017', 'Runtime': '133 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Jon Watts', 'Writer': 'Jonathan Goldstein (screenplay by), John Francis Daley (screenplay by), Jon Watts (screenplay by), Christopher Ford (screenplay by), Chris McKenna (screenplay by), Erik Sommers (screenplay by), Jonathan Goldstein (screen story by), John Francis Daley (screen story by), Stan Lee (based on the Marvel comic book by), Steve Ditko (based on the Marvel comic book by), Joe Simon (Captain America created by), Jack Kirby (Captain America crea

40
https://m.media-amazon.com/images/M/MV5BMTczNTI2ODUwOF5BMl5BanBnXkFtZTcwMTU0NTIzMw@@._V1_SX300.jpg
Movie title: Iron Man
Iron Man
0371746
Robert Downey Jr., Terrence Howard, Jeff Bridges, Gwyneth Paltrow
Robert Downey Jr. : Tony Stark : None
Second_cast actor: Terrence Howard : Rhodey
===end=====
{'Title': 'Avengers: Infinity War', 'Year': '2018', 'Rated': 'PG-13', 'Released': '27 Apr 2018', 'Runtime': '149 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Anthony Russo, Joe Russo', 'Writer': 'Christopher Markus (screenplay by), Stephen McFeely (screenplay by), Stan Lee (based on the Marvel comics by), Jack Kirby (based on the Marvel comics by), Joe Simon (Captain America created by), Jack Kirby (Captain America created by), Steve Englehart (Star-Lord created by), Steve Gan (Star-Lord created by), Bill Mantlo (Rocket Raccoon created by), Keith Giffen (Rocket Raccoon created by), Jim Starlin (Thanos,  Gamora and Drax created by), Stan Lee (Groot created by), Larry Lieber (Groo

45
https://m.media-amazon.com/images/M/MV5BMzA0YWMwMTUtMTVhNC00NjRkLWE2ZTgtOWEzNjJhYzNiMTlkXkEyXkFqcGdeQXVyNjc1NTYyMjg@._V1_SX300.jpg
Movie title: Superman
Superman
0078346
Marlon Brando, Gene Hackman, Christopher Reeve, Ned Beatty
Marlon Brando : Jor-El : None
Second_cast actor: Gene Hackman : Lex Luthor
===end=====
{'Title': 'The Avengers', 'Year': '2012', 'Rated': 'PG-13', 'Released': '04 May 2012', 'Runtime': '143 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Joss Whedon', 'Writer': 'Joss Whedon (screenplay), Zak Penn (story), Joss Whedon (story)', 'Actors': 'Robert Downey Jr., Chris Evans, Mark Ruffalo, Chris Hemsworth', 'Plot': "Earth's mightiest heroes must come together and learn to fight as a team if they are going to stop the mischievous Loki and his alien army from enslaving humanity.", 'Language': 'English, Russian, Hindi', 'Country': 'USA', 'Awards': 'Nominated for 1 Oscar. Another 38 wins & 79 nominations.', 'Poster': 'https://m.media-amazon.com/images/M/MV5BND

In [75]:
# Creating a dict to hold an info
about_movie_dict = {
        'imdbID':imdbID,
        'id_clean': id_clean,
        'Title': Title,
        'Actor': Actor,
        'Role1': Role1,
        'Role2': Role2,
        'Year': Year,
        'Production': Production,
        'Plot': Plot,
        'Poster': Poster
}

print(len(about_movie_dict['Title']))

50


In [77]:
# Creating a data frame about_movie_df
about_movie_df = pd.DataFrame(about_movie_dict)
print(about_movie_df.count())
# Saving data to csv file
about_movie_df.to_csv('about_movie.csv', index=False)
about_movie_df.head()

imdbID        50
id_clean      50
Title         50
Actor         50
Role1         50
Role2         50
Year          50
Production    50
Plot          50
Poster        50
dtype: int64


imdbID id_clean                        Title            Actor  \
0  tt0109506  0109506                     The Crow      Brandon Lee   
1  tt0112462  0112462               Batman Forever       Val Kilmer   
2  tt0411477  0411477  Hellboy II: The Golden Army      Ron Perlman   
3  tt1981115  1981115         Thor: The Dark World  Chris Hemsworth   
4  tt0770828  0770828                 Man of Steel     Henry Cavill   

        Role1        Role2  Year               Production  \
0        Eric         None  1994  LionsGate Entertainment   
1      Batman  Bruce Wayne  1995    Warner Bros. Pictures   
2     Hellboy         None  2008       Universal Pictures   
3        Thor         None  2013     Walt Disney Pictures   
4  Clark Kent       Kal-El  2013    Warner Bros. Pictures   

                                                Plot  \
0  A man brutally murdered comes back to life as ...   
1  Batman must battle former district attorney Ha...   
2  The mythical world starts a rebellion against ...   
3  When the Dark Elves attempt to plunge the univ...   
4  An alien child is evacuated from his dying wor...   

                                              Poster  
0  https://m.media-amazon.com/images/M/MV5BM2Y4ZG...  
1  https://m.media-amazon.com/images/M/MV5BNDdjYm...  
2  https://m.media-amazon.com/images/M/MV5BMjA5Nz...  
3  https://m.media-amazon.com/images/M/MV5BMTQyNz...  
4  https://m.media-amazon.com/images/M/MV5BMTk5OD...

In [78]:
# Data cleaning, checking Role1 and Role2 columns, some of the rows have Hero name and Hero full name mixed.
# Checking that all Heroes names in Role1 and Heroes full name in the Role2, if they have one.
about_movie_df.loc[about_movie_df['Role1'] == 'Bruce Wayne', 'Role1'] = 'Batman'
about_movie_df.loc[about_movie_df['Role1'] == 'Batman','Role2'] = 'Bruce Wayne'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Hellboy', 'Role1'] = 'Hellboy'
about_movie_df.loc[about_movie_df['Role1'] == 'Hellboy','Role2'] = 'Anung Un Rama'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Tony Stark', 'Role1'] = 'Iron Man'
about_movie_df.loc[about_movie_df['Role1'] == 'Iron Man','Role2'] = 'Tony Stark'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Eric', 'Role1'] = 'Crow'
about_movie_df.loc[about_movie_df['Role1'] == 'Crow','Role2'] = 'Eric Draven'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Clark Kent', 'Role1'] = 'Superman'
about_movie_df.loc[about_movie_df['Role1'] == 'Superman','Role2'] = 'Clark Kent'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Thor', 'Role1'] = 'Thor'
about_movie_df.loc[about_movie_df['Role1'] == 'Thor','Role2'] = 'Thor Odinson'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Shazam', 'Role1'] = 'Captain Marvel'
about_movie_df.loc[about_movie_df['Title'] == 'Shazam!','Role2'] = 'Billy Batson'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Logan', 'Role1'] = 'Wolverine'
about_movie_df.loc[about_movie_df['Role1'] == 'Wolverine','Role2'] = 'Logan'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Scott Lang', 'Role1'] = 'Ant-Man'
about_movie_df.loc[about_movie_df['Role1'] == 'Ant-Man','Role2'] = 'Hank Pym'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Charles Xavier', 'Role1'] = 'Professor X'
about_movie_df.loc[about_movie_df['Role1'] == 'Professor X','Role2'] = 'Charles Francis Xavier'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Arthur', 'Role1'] = 'Aquaman'
about_movie_df.loc[about_movie_df['Role1'] == 'Aquaman','Role2'] = 'Orin'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Dr. Stephen Strange', 'Role1'] = 'Doctor Strange'
about_movie_df.loc[about_movie_df['Role1'] == 'Doctor Strange','Role2'] = 'Stephen Strange'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Miles Morales', 'Role1'] = 'Spider-Man'
about_movie_df.loc[about_movie_df['Role1'] == 'Spider-Man','Role2'] = 'Peter Parker'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Lex Luthor', 'Role1'] = 'Lex Luthor'
about_movie_df.loc[about_movie_df['Role1'] == 'Lex Luthor','Role2'] = 'Lex Luthor'
about_movie_df


about_movie_df.loc[about_movie_df['Role1'] == 'Professor Charles Xavier', 'Role1'] = 'Professor X'
about_movie_df.loc[about_movie_df['Role1'] == 'Professor X','Role2'] = 'Charles Francis Xavier'
about_movie_df

about_movie_df.loc[about_movie_df['Actor'] == 'Ryan Reynolds', 'Role1'] = 'Juggernaut'
about_movie_df.loc[about_movie_df['Role1'] == 'Juggernaut','Role2'] = 'Cain Marko'
about_movie_df

about_movie_df.loc[about_movie_df['Actor'] == 'Brie Larson', 'Role1'] = 'Captain Marvel'
about_movie_df.loc[about_movie_df['Actor'] == 'Brie Larson','Role2'] = 'Carol Danvers'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Steve Rogers', 'Role1'] = 'Captain America'
about_movie_df.loc[about_movie_df['Role1'] == 'Captain America','Role2'] = 'Steve Rogers'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Peter Quill', 'Role1'] = 'Star-Lord'
about_movie_df.loc[about_movie_df['Role1'] == 'Star-Lord','Role2'] = 'Peter Jason Quill'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Peter Parker', 'Role1'] = 'Spider-Man'
about_movie_df.loc[about_movie_df['Role1'] == 'Spider-Man','Role2'] = 'Peter Parker'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == 'Diana', 'Role1'] = 'Wonder Woman'
about_movie_df.loc[about_movie_df['Role1'] == 'Wonder Woman','Role2'] = 'Diana Prince'
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == "Jor-El", 'Role1'] = 'Jor-El'
about_movie_df.loc[about_movie_df['Role1'] == 'Jor-El','Role2'] = "Jor-El"
about_movie_df

about_movie_df.loc[about_movie_df['Role1'] == "T'Challa", 'Role1'] = 'Black Panther'
about_movie_df.loc[about_movie_df['Role1'] == 'Black Panther','Role2'] = "T'Challa"
about_movie_df

about_movie_df.loc[about_movie_df['Title'] == "Spider-Man: Far from Home", 'Title'] = 'Spider-Man: Far From Home'

about_movie_df.to_csv('about_movie.csv', index=False)
about_movie_df

imdbID id_clean                                Title  \
0   tt0109506  0109506                             The Crow   
1   tt0112462  0112462                       Batman Forever   
2   tt0411477  0411477          Hellboy II: The Golden Army   
3   tt1981115  1981115                 Thor: The Dark World   
4   tt0770828  0770828                         Man of Steel   
5   tt0800369  0800369                                 Thor   
6   tt0448115  0448115                              Shazam!   
7   tt0376994  0376994                X-Men: The Last Stand   
8   tt0478970  0478970                              Ant-Man   
9   tt0948470  0948470               The Amazing Spider-Man   
10  tt0458339  0458339   Captain America: The First Avenger   
11  tt0348150  0348150                     Superman Returns   
12  tt1270798  1270798                   X-Men: First Class   
13  tt0120903  0120903                                X-Men   
14  tt5095030  5095030                 Ant-Man and the Wasp   
15  tt1477834  1477834                              Aquaman   
16  tt0413300  0413300                         Spider-Man 3   
17  tt1211837  1211837                       Doctor Strange   
18  tt0103776  0103776                       Batman Returns   
19  tt0372784  0372784                        Batman Begins   
20  tt1877832  1877832           X-Men: Days of Future Past   
21  tt4633694  4633694    Spider-Man: Into the Spider-Verse   
22  tt0081573  0081573                          Superman II   
23  tt0290334  0290334                     X2: X-Men United   
24  tt1228705  1228705                           Iron Man 2   
25  tt5463162  5463162                           Deadpool 2   
26  tt3315342  3315342                                Logan   
27  tt4154664  4154664                       Captain Marvel   
28  tt1843866  1843866  Captain America: The Winter Soldier   
29  tt2395427  2395427              Avengers: Age of Ultron   
30  tt3501632  3501632                       Thor: Ragnarok   
31  tt3896198  3896198       Guardians of the Galaxy Vol. 2   
32  tt0096895  0096895                               Batman   
33  tt1431045  1431045                             Deadpool   
34  tt1300854  1300854                           Iron Man 3   
35  tt6320628  6320628            Spider-Man: Far From Home   
36  tt2250912  2250912               Spider-Man: Homecoming   
37  tt2015381  2015381              Guardians of the Galaxy   
38  tt1345836  1345836                The Dark Knight Rises   
39  tt3498820  3498820           Captain America: Civil War   
40  tt0371746  0371746                             Iron Man   
41  tt4154756  4154756               Avengers: Infinity War   
42  tt0451279  0451279                         Wonder Woman   
43  tt0316654  0316654                         Spider-Man 2   
44  tt0145487  0145487                           Spider-Man   
45  tt0078346  0078346                             Superman   
46  tt0848228  0848228                         The Avengers   
47  tt4154796  4154796                    Avengers: Endgame   
48  tt1825683  1825683                        Black Panther   
49  tt0468569  0468569                      The Dark Knight   

                   Actor            Role1                   Role2  Year  \
0            Brandon Lee             Crow             Eric Draven  1994   
1             Val Kilmer           Batman             Bruce Wayne  1995   
2            Ron Perlman          Hellboy           Anung Un Rama  2008   
3        Chris Hemsworth             Thor            Thor Odinson  2013   
4           Henry Cavill         Superman              Clark Kent  2013   
5        Chris Hemsworth             Thor            Thor Odinson  2011   
6           Zachary Levi   Captain Marvel            Billy Batson  2019   
7           Hugh Jackman        Wolverine                   Logan  2006   
8              Paul Rudd          Ant-Man                Hank Pym  2015   
9        Andrew Garfield       Spider-Man  

In [79]:
# Adding a missing Hero - the Crow

the_crow_dict = {
        "name" : ['Crow'],
        "fullName":['Eric Draven'],
        "gender":['Male'],
        "eyeColor" :['Blue'],
        "height":['avr'],
        "weight":['avr'],
        "race" :['human'],
        "placeOfBirth": ['Earth'],
        "work" : ['a rock musician'],
        "images": ['https://upload.wikimedia.org/wikipedia/en/3/39/Crow_ver2.jpg'],
        "intelligence":['high'],
        "power":['high']
}



the_crow_df = pd.DataFrame(the_crow_dict)
print(the_crow_df.count())
the_crow_df.head()

name            1
fullName        1
gender          1
eyeColor        1
height          1
weight          1
race            1
placeOfBirth    1
work            1
images          1
intelligence    1
power           1
dtype: int64


name     fullName gender eyeColor height weight   race placeOfBirth  \
0  Crow  Eric Draven   Male     Blue    avr    avr  human        Earth   

              work                                             images  \
0  a rock musician  https://upload.wikimedia.org/wikipedia/en/3/39...   

  intelligence power  
0         high  high

In [80]:
# Appending the Crow info to the new_hero_df
new_hero_df = hero_df.append(the_crow_df)

# Checking, that a Crow hero is in dataset   
new_hero_df.loc[new_hero_df['name'] == 'Crow']
# new_hero_df

name     fullName gender eyeColor height weight   race placeOfBirth  \
0  Crow  Eric Draven   Male     Blue    avr    avr  human        Earth   

              work                                             images  \
0  a rock musician  https://upload.wikimedia.org/wikipedia/en/3/39...   

  intelligence power  
0         high  high

In [81]:
# Adding a missing Hero - the Jor-El
jor_el_dict = {
        "name" : ['Jor-El'],
        "fullName":['Jor-El'],
        "gender":['Male'],
        "eyeColor" :['Blue'],
        "height":['avr'],
        "weight":['avr'],
        "race" :['human'],
        "placeOfBirth": ['Krypton,'],
        "work" : ['a scientist'],
        "images": ['https://vignette.wikia.nocookie.net/supermanrebirth/images/3/3d/Marlon_Brando_Jor-El.jpg/revision/latest?cb=20130224095531'],
        "intelligence":['high'],
        "power":['high']
}



jor_el_df = pd.DataFrame(jor_el_dict)
print(jor_el_df.count())
jor_el_df.head()



name            1
fullName        1
gender          1
eyeColor        1
height          1
weight          1
race            1
placeOfBirth    1
work            1
images          1
intelligence    1
power           1
dtype: int64


name fullName gender eyeColor height weight   race placeOfBirth  \
0  Jor-El   Jor-El   Male     Blue    avr    avr  human     Krypton,   

          work                                             images  \
0  a scientist  https://vignette.wikia.nocookie.net/supermanre...   

  intelligence power  
0         high  high

In [82]:
# Adding missing hero Jor-El to new_hero_df data set
new_hero_df = new_hero_df.append(jor_el_df)

    
# new_df.loc[new_df['name'] == 'Jor-El']
new_hero_df.reset_index(drop = True)

name                fullName  gender eyeColor height  weight  \
0             A-Bomb  Richard Milhouse Jones    Male   Yellow    6'8  980 lb   
1         Abe Sapien          Abraham Sapien    Male     Blue    6'3  145 lb   
2           Abin Sur                            Male     Blue    6'1  200 lb   
3        Abomination            Emil Blonsky    Male    Green    6'8  980 lb   
4            Abraxas                 Abraxas    Male     Blue      -    - lb   
..               ...                     ...     ...      ...    ...     ...   
558  Yellowjacket II             Rita DeMara  Female     Blue    5'5  115 lb   
559             Ymir                    Ymir    Male    White   1000    - lb   
560             Yoda                    Yoda    Male    Brown    2'2   38 lb   
561             Crow             Eric Draven    Male     Blue    avr     avr   
562           Jor-El                  Jor-El    Male     Blue    avr     avr   

                  race        placeOfBirth  \
0                Human  Scarsdale, Arizona   
1        Icthyo Sapien                   -   
2              Ungaran              Ungara   
3    Human / Radiation  Zagreb, Yugoslavia   
4        Cosmic Entity     Within Eternity   
..                 ...                 ...   
558              Human                   -   
559        Frost Giant          Niffleheim   
560     Yoda's species                   -   
561              human               Earth   
562              human            Krypton,   

                                                  work  \
0    Musician, adventurer, author; formerly talk sh...   
1                              Paranormal Investigator   
2              Green Lantern, former history professor   
3                                               Ex-Spy   
4                                Dimensional destroyer   
..                                                 ...   
558  Adventurer; former criminal, electronics engineer   
559                                                  -   
560                                                  -   
561                                    a rock musician   
562                                        a scientist   

                                                images intelligence power  
0    https://cdn.rawgit.com/akabab/superhero-api/0....           38    24  
1    https://cdn.rawgit.com/akabab/superhero-api/0....           88   100  
2    https://cdn.rawgit.com/akabab/superhero-api/0....           50    99  
3    https://cdn.rawgit.com/akabab/superhero-api/0....           63    62  
4    https://cdn.rawgit.com/akabab/superhero-api/0....           88   100  
..                                                 ...          ...   ...  
558  https://cdn.rawgit.com/akabab/superhero-api/0....           50    31  
559  https://cdn.rawgit.com/akabab/superhero-api/0....           50    98  
560  https://cdn.rawgit.com/akabab/superhero-api/0....           88   100  
561  https://upload.wikimedia.org/wikipedia/en/3/39...         high  high  
562  https://vignette.wikia.nocookie.net/supermanre...         high  high  

[563 rows x 12 columns]

In [83]:
# Checking wich Heroes appeared the most in the top 50 SuperHero Movies
group_hero = about_movie_df.groupby(['Role1', 'Role2'])['Role1'].count()
group_hero_df = pd.DataFrame(group_hero)
group_hero_df = group_hero_df.rename(columns={'Role1':'Count'})
group_hero_df = group_hero_df.sort_values(by=['Count'], ascending=False)
group_hero_df
# Iron Man and Spider-Man appeared the most.

Count
Role1           Role2                        
Iron Man        Tony Stark                  7
Spider-Man      Peter Parker                7
Batman          Bruce Wayne                 6
Wolverine       Logan                       4
Captain America Steve Rogers                3
Thor            Thor Odinson                3
Juggernaut      Cain Marko                  2
Superman        Clark Kent                  2
Star-Lord       Peter Jason Quill           2
Professor X     Charles Francis Xavier      2
Ant-Man         Hank Pym                    2
Jor-El          Jor-El                      1
Aquaman         Orin                        1
Lex Luthor      Lex Luthor                  1
Hellboy         Anung Un Rama               1
Doctor Strange  Stephen Strange             1
Crow            Eric Draven                 1
Captain Marvel  Carol Danvers               1
                Billy Batson                1
Black Panther   T'Challa                    1
Wonder Woman    Diana Prince                1

In [84]:
# Joining about_movie_df and movies_df on Title and title 
# Group by Hero, to see wich Hero brings the best revenue.
movie_actor_role_df = pd.merge(left=about_movie_df, right=movies_df, left_on='Title', right_on='title')
movie_actor_role_df

imdbID id_clean                                Title  \
0   tt0109506  0109506                             The Crow   
1   tt0112462  0112462                       Batman Forever   
2   tt0411477  0411477          Hellboy II: The Golden Army   
3   tt1981115  1981115                 Thor: The Dark World   
4   tt0770828  0770828                         Man of Steel   
5   tt0800369  0800369                                 Thor   
6   tt0448115  0448115                              Shazam!   
7   tt0376994  0376994                X-Men: The Last Stand   
8   tt0478970  0478970                              Ant-Man   
9   tt0948470  0948470               The Amazing Spider-Man   
10  tt0458339  0458339   Captain America: The First Avenger   
11  tt0348150  0348150                     Superman Returns   
12  tt1270798  1270798                   X-Men: First Class   
13  tt0120903  0120903                                X-Men   
14  tt5095030  5095030                 Ant-Man and the Wasp   
15  tt1477834  1477834                              Aquaman   
16  tt0413300  0413300                         Spider-Man 3   
17  tt1211837  1211837                       Doctor Strange   
18  tt0103776  0103776                       Batman Returns   
19  tt0372784  0372784                        Batman Begins   
20  tt1877832  1877832           X-Men: Days of Future Past   
21  tt4633694  4633694    Spider-Man: Into the Spider-Verse   
22  tt0081573  0081573                          Superman II   
23  tt0290334  0290334                     X2: X-Men United   
24  tt1228705  1228705                           Iron Man 2   
25  tt5463162  5463162                           Deadpool 2   
26  tt3315342  3315342                                Logan   
27  tt4154664  4154664                       Captain Marvel   
28  tt1843866  1843866  Captain America: The Winter Soldier   
29  tt2395427  2395427              Avengers: Age of Ultron   
30  tt3501632  3501632                       Thor: Ragnarok   
31  tt3896198  3896198       Guardians of the Galaxy Vol. 2   
32  tt0096895  0096895                               Batman   
33  tt1431045  1431045                             Deadpool   
34  tt1300854  1300854                           Iron Man 3   
35  tt6320628  6320628            Spider-Man: Far From Home   
36  tt2250912  2250912               Spider-Man: Homecoming   
37  tt2015381  2015381              Guardians of the Galaxy   
38  tt1345836  1345836                The Dark Knight Rises   
39  tt3498820  3498820           Captain America: Civil War   
40  tt0371746  0371746                             Iron Man   
41  tt4154756  4154756               Avengers: Infinity War   
42  tt0451279  0451279                         Wonder Woman   
43  tt0316654  0316654                         Spider-Man 2   
44  tt0145487  0145487                           Spider-Man   
45  tt0078346  0078346                             Superman   
46  tt0848228  0848228                         The Avengers   
47  tt4154796  4154796                    Avengers: Endgame   
48  tt1825683  1825683                        Black Panther   
49  tt0468569  0468569                      The Dark Knight   

                   Actor            Role1                   Role2  Year  \
0            Brandon Lee             Crow             Eric Draven  1994   
1             Val Kilmer           Batman             Bruce Wayne  1995   
2            Ron Perlman          Hellboy           Anung Un Rama  2008   
3        Chris Hemsworth             Thor            Thor Odinson  2013   
4           Henry Cavill         Superman              Clark Kent  2013   
5        Chris Hemsworth             Thor            Thor Odinson  2011   
6           Zachary Levi   Captain Marvel            Billy Batson  2019   
7           Hugh Jackman        Wolverine                   Logan  2006   
8              Paul Rudd          Ant-Man                Hank Pym  2015   
9        Andrew Garfield       Spider-Man  

In [85]:
# Delete duplicate columns
movie_actor_role_df = movie_actor_role_df.drop(['id_clean', 'date', 'rewatchability', 'rottenTomatoes', 'totalScore', 'image'] , axis=1)
movie_actor_role_df.head()



imdbID                        Title            Actor     Role1  \
0  tt0109506                     The Crow      Brandon Lee      Crow   
1  tt0112462               Batman Forever       Val Kilmer    Batman   
2  tt0411477  Hellboy II: The Golden Army      Ron Perlman   Hellboy   
3  tt1981115         Thor: The Dark World  Chris Hemsworth      Thor   
4  tt0770828                 Man of Steel     Henry Cavill  Superman   

           Role2  Year               Production  \
0    Eric Draven  1994  LionsGate Entertainment   
1    Bruce Wayne  1995    Warner Bros. Pictures   
2  Anung Un Rama  2008       Universal Pictures   
3   Thor Odinson  2013     Walt Disney Pictures   
4     Clark Kent  2013    Warner Bros. Pictures   

                                                Plot  \
0  A man brutally murdered comes back to life as ...   
1  Batman must battle former district attorney Ha...   
2  The mythical world starts a rebellion against ...   
3  When the Dark Elves attempt to plunge the univ...   
4  An alien child is evacuated from his dying wor...   

                                              Poster rank  \
0  https://m.media-amazon.com/images/M/MV5BM2Y4ZG...   50   
1  https://m.media-amazon.com/images/M/MV5BNDdjYm...   49   
2  https://m.media-amazon.com/images/M/MV5BMjA5Nz...   48   
3  https://m.media-amazon.com/images/M/MV5BMTQyNz...   47   
4  https://m.media-amazon.com/images/M/MV5BMTk5OD...   46   

                         title  boxOffice  
0                     The Crow       83.3  
1               Batman Forever      295.0  
2  Hellboy II: The Golden Army       87.0  
3         Thor: The Dark World      219.0  
4                 Man of Steel      309.0

In [86]:
# Group by hero to find the top box office
group_hero_revenue = movie_actor_role_df.groupby('Role1')['boxOffice'].sum()
group_hero_revenue_df = pd.DataFrame(group_hero_revenue)
group_hero_revenue_df = group_hero_revenue_df.sort_values(by=['boxOffice'], ascending=False)
group_hero_revenue_df= group_hero_revenue_df.reset_index()
group_hero_revenue_df
# Iron Man is the most valuable hero.

Role1  boxOffice
0          Iron Man     3826.0
1        Spider-Man     2610.0
2            Batman     2421.0
3         Wolverine      979.0
4   Captain America      873.0
5         Star-Lord      737.0
6              Thor      733.0
7     Black Panther      700.0
8        Juggernaut      688.0
9          Superman      595.0
10   Captain Marvel      567.4
11           Jor-El      502.0
12      Professor X      448.0
13     Wonder Woman      412.0
14          Ant-Man      400.0
15          Aquaman      335.0
16       Lex Luthor      290.0
17   Doctor Strange      233.0
18          Hellboy       87.0
19             Crow       83.3

In [88]:
# Group by production to find the most valuable company
group_production_revenue = movie_actor_role_df.groupby('Production')['boxOffice'].sum()
group_production_revenue = pd.DataFrame(group_production_revenue)
group_production_revenue = group_production_revenue.sort_values(by=['boxOffice'], ascending=False)
group_production_revenue= group_production_revenue.reset_index()
group_production_revenue
# Walt Disney Pictures is the most profitable company

Production  boxOffice
0              Walt Disney Pictures     4651.0
1             Warner Bros. Pictures     3943.0
2                  20th Century Fox     2115.0
3                     Sony Pictures     2060.0
4                    Marvel Studios     1985.0
5                Paramount Pictures      757.0
6   Warner Bros. Pictures/Legendary      612.0
7                 Columbia Pictures      550.0
8                 Paramount Studios      353.0
9                     Disney/Marvel      183.0
10                  New Line Cinema      140.4
11               Universal Pictures       87.0
12          LionsGate Entertainment       83.3

In [87]:
#Bar chart for each Hero and revenue
import plotly.express as px
df = group_hero_revenue_df
fig = px.bar(df, x='Role1', y='boxOffice')

fig.update_xaxes(title_text='Hero')
fig.update_yaxes(title_text='Box Office ($million)')

fig.update_layout(
    title={
        'text':'Box Office per Hero',
        'y':0.95,
        'x':0.4,
        'xanchor': 'center',
        'yanchor': 'top'}
)


fig.show()

In [93]:
# Checking wich production company appeared the most in the top 50 SuperHero Movies
group_production = about_movie_df.groupby('Production')['Production'].count()
group_production_df = pd.DataFrame(group_production)
group_production_df = group_production_df.rename(columns={'Production':'Count'})
group_production_df = group_production_df.sort_values(by=['Count'], ascending=False)
group_production_df

Count
Production                            
Walt Disney Pictures                12
Warner Bros. Pictures               11
20th Century Fox                     8
Sony Pictures                        6
Marvel Studios                       3
Paramount Pictures                   3
Columbia Pictures                    1
Disney/Marvel                        1
LionsGate Entertainment              1
New Line Cinema                      1
Paramount Studios                    1
Universal Pictures                   1
Warner Bros. Pictures/Legendary      1

In [94]:
#Bar chart for each production studio vs revenue
import plotly.express as px
df = group_production_revenue
fig = px.bar(df, x='Production', y='boxOffice')

fig.update_xaxes(title_text='Production Studio')
fig.update_yaxes(title_text='Box Office ($million)')

fig.update_layout(
    title={
        'text':'Box Office per Production Studio',
        'y':0.95,
        'x':0.4,
        'xanchor': 'center',
        'yanchor': 'top'}
)


fig.show()

#### Data Analysis (Question 3)

1. Based on the data presented above we can see that the most valuable superhero is Iron Man with the total box office of 3826 million dollars. Iron man was in 7 movies, and they represent 14% of all our movies. 
2. The most profitable company is Walt Disney Pictures. It created 12 movies (24% of all our movies) with the total box office of 4651 million dollars. 
3. Base on this results, Walt Disney will keep producing superhero movies and traders will be bullish on Walt Disney stock, it means that they thinks the price of Walt Disney stock is going to go up. 

In [95]:
# PART 4 =========================== Final ======================
# Joining  movie_actor_role_df with new_hero_df on Role1 and name. 
# It will create final dataset, which will have info about the best 50 Superheroes movies, main actor for each movie 
# and the main hero for each movie.

movie_actor_hero_df = pd.merge(left=movie_actor_role_df, right=new_hero_df, left_on=['Role1', 'Role2'], right_on=['name', 'fullName'])
movie_actor_hero_df



imdbID                                Title                 Actor  \
0   tt0109506                             The Crow           Brandon Lee   
1   tt0112462                       Batman Forever            Val Kilmer   
2   tt0103776                       Batman Returns        Michael Keaton   
3   tt0372784                        Batman Begins        Christian Bale   
4   tt0096895                               Batman        Michael Keaton   
5   tt1345836                The Dark Knight Rises        Christian Bale   
6   tt0468569                      The Dark Knight        Christian Bale   
7   tt0411477          Hellboy II: The Golden Army           Ron Perlman   
8   tt1981115                 Thor: The Dark World       Chris Hemsworth   
9   tt0800369                                 Thor       Chris Hemsworth   
10  tt3501632                       Thor: Ragnarok       Chris Hemsworth   
11  tt0770828                         Man of Steel          Henry Cavill   
12  tt0348150                     Superman Returns         Brandon Routh   
13  tt0448115                              Shazam!          Zachary Levi   
14  tt0376994                X-Men: The Last Stand          Hugh Jackman   
15  tt0120903                                X-Men          Hugh Jackman   
16  tt1877832           X-Men: Days of Future Past          Hugh Jackman   
17  tt3315342                                Logan          Hugh Jackman   
18  tt0478970                              Ant-Man             Paul Rudd   
19  tt5095030                 Ant-Man and the Wasp             Paul Rudd   
20  tt0948470               The Amazing Spider-Man       Andrew Garfield   
21  tt0413300                         Spider-Man 3         Tobey Maguire   
22  tt4633694    Spider-Man: Into the Spider-Verse         Shameik Moore   
23  tt6320628            Spider-Man: Far From Home           Tom Holland   
24  tt2250912               Spider-Man: Homecoming           Tom Holland   
25  tt0316654                         Spider-Man 2         Tobey Maguire   
26  tt0145487                           Spider-Man         Tobey Maguire   
27  tt0458339   Captain America: The First Avenger           Chris Evans   
28  tt1843866  Captain America: The Winter Soldier           Chris Evans   
29  tt3498820           Captain America: Civil War           Chris Evans   
30  tt1270798                   X-Men: First Class          James McAvoy   
31  tt0290334                     X2: X-Men United       Patrick Stewart   
32  tt1477834                              Aquaman           Jason Momoa   
33  tt1211837                       Doctor Strange  Benedict Cumberbatch   
34  tt0081573                          Superman II          Gene Hackman   
35  tt1228705                           Iron Man 2     Robert Downey Jr.   
36  tt2395427              Avengers: Age of Ultron     Robert Downey Jr.   
37  tt1300854                           Iron Man 3     Robert Downey Jr.   
38  tt0371746                             Iron Man     Robert Downey Jr.   
39  tt4154756               Avengers: Infinity War     Robert Downey Jr.   
40  tt0848228                         The Avengers     Robert Downey Jr.   
41  tt4154796                    Avengers: Endgame     Robert Downey Jr.   
42  tt5463162                           Deadpool 2         Ryan Reynolds   
43  tt1431045                             Deadpool         Ryan Reynolds   
44  tt4154664                       Captain Marvel           Brie Larson   
45  tt3896198       Guardians of the Galaxy Vol. 2           Chris Pratt   
46  tt2015381              Guardians of the Galaxy           Chris Pratt   
47  tt0451279                         Wonder Woman             Gal Gadot   
48  tt0078346                             Superman         Marlon Brando   
49  tt1825683                        Black Panther      Chadwick Boseman   

              Role1                   Role2  Year  \
0              Crow             Eric Draven  1994   
1            Batman     

In [96]:
# Cleaning data for movie_actor_hero_df
movie_actor_hero_df = movie_actor_hero_df.drop(['imdbID','title', 'Role1' , 'Role2', 'Production', 'eyeColor', 'height', 'weight', 'race', 'intelligence', 'power'], axis = 1)



In [97]:
movie_actor_hero_df.head(1)


Title        Actor  Year  \
0  The Crow  Brandon Lee  1994   

                                                Plot  \
0  A man brutally murdered comes back to life as ...   

                                              Poster rank  boxOffice  name  \
0  https://m.media-amazon.com/images/M/MV5BM2Y4ZG...   50       83.3  Crow   

      fullName gender placeOfBirth             work  \
0  Eric Draven   Male        Earth  a rock musician   

                                              images  
0  https://upload.wikimedia.org/wikipedia/en/3/39...

In [98]:
# Renaming columns
movie_actor_hero_df = movie_actor_hero_df.rename(columns={"rank": "Movie_Rank", 
                                                        "boxOffice": "Box_Office",
                                                        "name": "Hero_Name",
                                                        "fullName": "Hero_Full_Name",
                                                        "gender": "Hero_Genger",
                                                        "placeOfBirth": "Hero_Place_Of_Birth",
                                                        "work": "Hero_work",
                                                        "images": "Hero_Images"
                                                       })

movie_actor_hero_df

Title                 Actor  Year  \
0                              The Crow           Brandon Lee  1994   
1                        Batman Forever            Val Kilmer  1995   
2                        Batman Returns        Michael Keaton  1992   
3                         Batman Begins        Christian Bale  2005   
4                                Batman        Michael Keaton  1989   
5                 The Dark Knight Rises        Christian Bale  2012   
6                       The Dark Knight        Christian Bale  2008   
7           Hellboy II: The Golden Army           Ron Perlman  2008   
8                  Thor: The Dark World       Chris Hemsworth  2013   
9                                  Thor       Chris Hemsworth  2011   
10                       Thor: Ragnarok       Chris Hemsworth  2017   
11                         Man of Steel          Henry Cavill  2013   
12                     Superman Returns         Brandon Routh  2006   
13                              Shazam!          Zachary Levi  2019   
14                X-Men: The Last Stand          Hugh Jackman  2006   
15                                X-Men          Hugh Jackman  2000   
16           X-Men: Days of Future Past          Hugh Jackman  2014   
17                                Logan          Hugh Jackman  2017   
18                              Ant-Man             Paul Rudd  2015   
19                 Ant-Man and the Wasp             Paul Rudd  2018   
20               The Amazing Spider-Man       Andrew Garfield  2012   
21                         Spider-Man 3         Tobey Maguire  2007   
22    Spider-Man: Into the Spider-Verse         Shameik Moore  2018   
23            Spider-Man: Far From Home           Tom Holland  2019   
24               Spider-Man: Homecoming           Tom Holland  2017   
25                         Spider-Man 2         Tobey Maguire  2004   
26                           Spider-Man         Tobey Maguire  2002   
27   Captain America: The First Avenger           Chris Evans  2011   
28  Captain America: The Winter Soldier           Chris Evans  2014   
29           Captain America: Civil War           Chris Evans  2016   
30                   X-Men: First Class          James McAvoy  2011   
31                     X2: X-Men United       Patrick Stewart  2003   
32                              Aquaman           Jason Momoa  2018   
33                       Doctor Strange  Benedict Cumberbatch  2016   
34                          Superman II          Gene Hackman  1980   
35                           Iron Man 2     Robert Downey Jr.  2010   
36              Avengers: Age of Ultron     Robert Downey Jr.  2015   
37                           Iron Man 3     Robert Downey Jr.  2013   
38                             Iron Man     Robert Downey Jr.  2008   
39               Avengers: Infinity War     Robert Downey Jr.  2018   
40                         The Avengers     Robert Downey Jr.  2012   
41                    Avengers: Endgame     Robert Downey Jr.  2019   
42                           Deadpool 2         Ryan Reynolds  2018   
43                             Deadpool         Ryan Reynolds  2016   
44                       Captain Marvel           Brie Larson  2019   
45       Guardians of the Galaxy Vol. 2           Chris Pratt  2017   
46              Guardians of the Galaxy           Chris Pratt  2014   
47                         Wonder Woman             Gal Gadot  2017   
48                             Superman         Marlon Brando  1978   
49                        Black Panther      Chadwick Boseman  2018   

                                                 Plot  \
0   A man brutally murdered comes back to life as ...   
1   Batman must battle former district attorney Ha...   
2   Batman returns to the big screen when a deform...   
3   After training with his mentor, Batman begins ...   
4   The Dark Knight of Gotham City begins his war ...   
5   Eight years after the Joker's reign of anarchy...   
6   When the 

In [99]:
# saving the dataframe to movie_actor_hero.csv file
movie_actor_hero_df.to_csv('movie_actor_hero.csv', index=False) 
movie_actor_hero_df.dtypes

Title                   object
Actor                   object
Year                    object
Plot                    object
Poster                  object
Movie_Rank              object
Box_Office             float64
Hero_Name               object
Hero_Full_Name          object
Hero_Genger             object
Hero_Place_Of_Birth     object
Hero_work               object
Hero_Images             object
dtype: object

In [1]:
from sqlalchemy import create_engine, Column, Integer, String
from sqlalchemy.ext.declarative import declarative_base
Base = declarative_base()
import psycopg2

# Define movie_actor_hero table
class movie_actor_hero(Base):
    __tablename__ = 'movie_actor_hero'
    id = Column(Integer, primary_key=True)
    Title=Column(String)
    Actor=Column(String)
    Year=Column(String)
    Plot=Column(String)
    Poster=Column(String)
    Movie_Rank=Column(String)
    Box_Office =Column(Integer)
    Hero_Name=Column(String)
    Hero_Full_Name=Column(String)
    Hero_Genger=Column(String)
    Hero_Place_Of_Birth=Column(String)
    Hero_work=Column(String)
    Hero_Images=Column(String)

In [2]:
Base.metadata.tables

immutabledict({'movie_actor_hero': Table('movie_actor_hero', MetaData(bind=None), Column('id', Integer(), table=<movie_actor_hero>, primary_key=True, nullable=False), Column('Title', String(), table=<movie_actor_hero>), Column('Actor', String(), table=<movie_actor_hero>), Column('Year', String(), table=<movie_actor_hero>), Column('Plot', String(), table=<movie_actor_hero>), Column('Poster', String(), table=<movie_actor_hero>), Column('Movie_Rank', String(), table=<movie_actor_hero>), Column('Box_Office', Integer(), table=<movie_actor_hero>), Column('Hero_Name', String(), table=<movie_actor_hero>), Column('Hero_Full_Name', String(), table=<movie_actor_hero>), Column('Hero_Genger', String(), table=<movie_actor_hero>), Column('Hero_Place_Of_Birth', String(), table=<movie_actor_hero>), Column('Hero_work', String(), table=<movie_actor_hero>), Column('Hero_Images', String(), table=<movie_actor_hero>), schema=None)})

In [5]:
# Transform premise DataFrame
csv_file = "movie_actor_hero.csv"
premise_data_df = pd.read_csv(csv_file)
premise_data_df.head()

Title           Actor  Year  \
0        The Crow     Brandon Lee  1994   
1  Batman Forever      Val Kilmer  1995   
2  Batman Returns  Michael Keaton  1992   
3   Batman Begins  Christian Bale  2005   
4          Batman  Michael Keaton  1989   

                                                Plot  \
0  A man brutally murdered comes back to life as ...   
1  Batman must battle former district attorney Ha...   
2  Batman returns to the big screen when a deform...   
3  After training with his mentor, Batman begins ...   
4  The Dark Knight of Gotham City begins his war ...   

                                              Poster  Movie_Rank  Box_Office  \
0  https://m.media-amazon.com/images/M/MV5BM2Y4ZG...          50        83.3   
1  https://m.media-amazon.com/images/M/MV5BNDdjYm...          49       295.0   
2  https://m.media-amazon.com/images/M/MV5BOGZmYz...          32       283.0   
3  https://m.media-amazon.com/images/M/MV5BZmUwNG...          31       256.0   
4  https://m.media-amazon.com/images/M/MV5BMTYwNj...          18       493.0   

  Hero_Name Hero_Full_Name Hero_Genger  \
0      Crow    Eric Draven        Male   
1    Batman    Bruce Wayne        Male   
2    Batman    Bruce Wayne        Male   
3    Batman    Bruce Wayne        Male   
4    Batman    Bruce Wayne        Male   

                           Hero_Place_Of_Birth        Hero_work  \
0                                        Earth  a rock musician   
1  Crest Hill, Bristol Township; Gotham County      Businessman   
2  Crest Hill, Bristol Township; Gotham County      Businessman   
3  Crest Hill, Bristol Township; Gotham County      Businessman   
4  Crest Hill, Bristol Township; Gotham County      Businessman   

                                         Hero_Images  
0  https://upload.wikimedia.org/wikipedia/en/3/39...  
1  https://cdn.rawgit.com/akabab/superhero-api/0....  
2  https://cdn.rawgit.com/akabab/superhero-api/0....  
3  https://cdn.rawgit.com/akabab/superhero-api/0....  
4  https://cdn.rawgit.com/akabab/superhero-api/0....

In [9]:
# Creating a DataBase
engine=psycopg2.connect(f'postgresql://postgres:postgres@localhost:5432')
engine.autocommit=True
cursor=engine.cursor()
sql='''Create database best_50_movies_heroes_db''';
cursor.execute(sql)

In [10]:
# Create database connection
engine=create_engine(f'postgresql://postgres:postgres@localhost:5432/best_50_movies_heroes_db')

In [11]:
# Confirm tables
Base.metadata.create_all(engine)
engine.table_names()

['movie_actor_hero']

In [12]:
# Load DataFrames into database
premise_data_df.to_sql(name='movie_actor_hero', con=engine, if_exists='append', index=False)

In [13]:
# Confirm data has been added by querying the movie_actor_hero table
pd.read_sql_query('select * from movie_actor_hero', con=engine).head()

id           Title           Actor  Year  \
0   1        The Crow     Brandon Lee  1994   
1   2  Batman Forever      Val Kilmer  1995   
2   3  Batman Returns  Michael Keaton  1992   
3   4   Batman Begins  Christian Bale  2005   
4   5          Batman  Michael Keaton  1989   

                                                Plot  \
0  A man brutally murdered comes back to life as ...   
1  Batman must battle former district attorney Ha...   
2  Batman returns to the big screen when a deform...   
3  After training with his mentor, Batman begins ...   
4  The Dark Knight of Gotham City begins his war ...   

                                              Poster Movie_Rank  Box_Office  \
0  https://m.media-amazon.com/images/M/MV5BM2Y4ZG...         50          83   
1  https://m.media-amazon.com/images/M/MV5BNDdjYm...         49         295   
2  https://m.media-amazon.com/images/M/MV5BOGZmYz...         32         283   
3  https://m.media-amazon.com/images/M/MV5BZmUwNG...         31         256   
4  https://m.media-amazon.com/images/M/MV5BMTYwNj...         18         493   

  Hero_Name Hero_Full_Name Hero_Genger  \
0      Crow    Eric Draven        Male   
1    Batman    Bruce Wayne        Male   
2    Batman    Bruce Wayne        Male   
3    Batman    Bruce Wayne        Male   
4    Batman    Bruce Wayne        Male   

                           Hero_Place_Of_Birth        Hero_work  \
0                                        Earth  a rock musician   
1  Crest Hill, Bristol Township; Gotham County      Businessman   
2  Crest Hill, Bristol Township; Gotham County      Businessman   
3  Crest Hill, Bristol Township; Gotham County      Businessman   
4  Crest Hill, Bristol Township; Gotham County      Businessman   

                                         Hero_Images  
0  https://upload.wikimedia.org/wikipedia/en/3/39...  
1  https://cdn.rawgit.com/akabab/superhero-api/0....  
2  https://cdn.rawgit.com/akabab/superhero-api/0....  
3  https://cdn.rawgit.com/akabab/superhero-api/0....  
4  https://cdn.rawgit.com/akabab/superhero-api/0....